diff -up nmap-5.51/ncat/ncat_listen.c.udpfix nmap-5.51/ncat/ncat_listen.c
--- nmap-5.51/ncat/ncat_listen.c.udpfix	2011-01-21 01:04:51.000000000 +0100
+++ nmap-5.51/ncat/ncat_listen.c	2013-05-21 15:37:51.396358761 +0200
@@ -497,7 +497,8 @@ static int ncat_listen_dgram(int proto)
             }
 
             /* Dump the current datagram */
-            Recv(sockfd, buf, sizeof(buf), 0);
+            nbytes = Recv(sockfd, buf, sizeof(buf), 0);
+            ncat_log_recv(buf, nbytes);
         }
 
         conn_inc++;
@@ -553,6 +554,7 @@ static int ncat_listen_dgram(int proto)
                         send(sockfd, tempbuf, nbytes, 0);
                     else
                         send(sockfd, buf, nbytes, 0);
+		    ncat_log_send(buf, nbytes);
                 }
                 if (tempbuf != NULL) {
                     free(tempbuf);
@@ -566,6 +568,7 @@ static int ncat_listen_dgram(int proto)
                     close(sockfd);
                     return 1;
                 }
+                ncat_log_recv(buf, nbytes);
                 if (!o.sendonly)
                     Write(STDOUT_FILENO, buf, nbytes);
             }
